misc/loop/cmd/snapshotter.go

```diff
@@ -0,0 +1,243 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/gnolang/tx-archive/backup"
+	"github.com/gnolang/tx-archive/backup/client/http"
+	"github.com/gnolang/tx-archive/backup/writer/legacy"
+)
+
+const (
+	setReadOnly   = `middlewares: ["ipwhitelist"]`
+	unsetReadOnly = `middlewares: []`
+)
+
+type snapshotter struct {
+	dockerClient *client.Client
+
+	containerName      string
+	backupFile         string
+	instanceBackupFile string
+
+	cfg config
+
+	url string
+}
+
+type config struct {
+	rpcAddr        string
+	traefikGnoFile string
+	backupDir      string
+	hostPWD        string
+}
+
+func NewSnapshotter(dockerClient *client.Client, cfg config) (*snapshotter, error) {
+	timenow := time.Now()
+	now := fmt.Sprintf("%s_%v", timenow.Format("2006-01-02_"), timenow.UnixNano())
+
+	backupFile, err := filepath.Abs(cfg.backupDir + "/backup.jsonl")
+	if err != nil {
+		return nil, err
+	}
+	instanceBackupFile, err := filepath.Abs(fmt.Sprintf("%s/backup_%s.jsonl", cfg.backupDir, now))
+	if err != nil {
+		return nil, err
+	}
+	return &snapshotter{
+		dockerClient: dockerClient,
+
+		cfg: cfg,
+
+		containerName:      "gno-" + now,
+		backupFile:         backupFile,
+		instanceBackupFile: instanceBackupFile,
+	}, nil
+}
+
+// pullLatestImage get latest version of the docker image
+func (s snapshotter) pullLatestImage(ctx context.Context) (bool, error) {
+	reader, err := s.dockerClient.ImagePull(ctx, "ghcr.io/gnolang/gno", types.ImagePullOptions{})
+	if err != nil {
+		return false, err
+	}
+	var b bytes.Buffer
+	defer reader.Close()
+
+	_, err = io.Copy(&b, reader)
+	if err != nil {
+		return false, err
+	}
+
+	return !strings.Contains(b.String(), "Image is up to date"), nil
+}
+
+func (s snapshotter) switchTraefikMode(replaceStr string) error {
+	input, err := ioutil.ReadFile(s.cfg.traefikGnoFile)
+	if err != nil {
+		return err
+	}
+
+	regex := regexp.MustCompile(`middlewares: \[.*\]`)
+	output := regex.ReplaceAllLiteral(input, []byte(replaceStr))
+
+	return ioutil.WriteFile(s.cfg.traefikGnoFile, output, 0655)
+}
+
+func (s snapshotter) switchTraefikPortalLoop(url string) error {
+	input, err := ioutil.ReadFile(s.cfg.traefikGnoFile)
+	if err != nil {
+		return err
+	}
+
+	regex := regexp.MustCompile(`http://.*:[0-9]+`)
+	output := regex.ReplaceAllLiteral(input, []byte(url))
```

Comment:
A hack, but it works. One thing to note though...

This is modifying the file in a volume, modifying directly the files tracked by git. This has an implication: if the code updates on the traefik configuration on Git, `git pull` will fail if portalloopd has modified the traefik config file.

Suggestion: maybe make `gno.yml` a template which is modified by `portalloopd`, so we have a `gno.yml.template` (tracked with git) and a `gno.yml` (untracked)...